src/ast/passes/semantic_analyser.cpp

```diff
@@ -769,6 +769,31 @@ void SemanticAnalyser::visit(Call &call)
       }
     }
     call.type = CreateNone();
+  } else if (call.func == "has_key") {
+    check_assignment(call, false, false, false);
+    if (check_varargs(call, 2, 2)) {
+      auto &arg0 = *call.vargs.at(0);
+      if (!arg0.is_map)
+        LOG(ERROR, arg0.loc, err_)
+            << "has_key() expects the first argument to be a map";
+
+      Map &map = static_cast<Map &>(arg0);
+      auto *mapkey = get_map_key_type(map);
+      if (mapkey) {
+        if (mapkey->IsNoneTy()) {
+          LOG(ERROR, arg0.loc, err_)
+              << "has_key() only accepts maps that have keys to iterate over";
+        } else {
+          auto &arg1 = *call.vargs.at(1);
+          SizedType new_key_type = create_key_type(arg1.type, arg1.loc);
+          update_key_type(*mapkey, new_key_type, map.ident, arg1.loc);
```

Comment:
Do we really want to update the stored key type based on the argument used in the `has_map` call?

Reply:
We need to do this for string resizing e.g. we don't want this to fail `@a["hi"] = 1; if (has_key(@a, "longerstr"))`

Reply:
Ok, makes sense.